src/main/scala/util/RunPeriodically.scala

```diff
@@ -0,0 +1,34 @@
+package lila.ws
+package util
+
+import org.apache.pekko.actor.Cancellable
+import org.apache.pekko.actor.typed.Scheduler
+
+/** Run a function after a counter hits a ceiling, or periodically after a fixed time interval, if counter > 0
+  */
+final class RunPeriodically()(using Scheduler, Executor):
+  def apply(counterMax: Int, interval: FiniteDuration)(flush: () => Unit) =
+    RunPeriodicallyStage(counterMax, interval, flush)
+
+final class RunPeriodicallyStage(counterMax: Int, interval: FiniteDuration, run: () => Unit)(using
+    scheduler: Scheduler,
+    ec: Executor
+):
+
+  private var counter: Int = 0
+
+  private var scheduledFlush: Cancellable = scheduler.scheduleOnce(interval, () => run())
+
+  def increment(): Unit =
+    synchronized:
+      counter += 1
+      if counter >= counterMax then unsafeFlush()
+
+  private def flush(): Unit = synchronized { unsafeFlush() }
+
+  private def unsafeFlush(): Unit =
+    if counter > 0 then
+      run()
+      counter == 0
```

Comment:
should it be `counter = 0` 🤔

Reply:
Dammit discarded values!!

Reply:
i don't really understand what this does. i was thinking something like:

```scala
package lila.ws
package util

import io.netty.channel.*
import org.apache.pekko.actor.Cancellable
import org.apache.pekko.actor.typed.Scheduler

final class RunPeriodically()(using scheduler: Scheduler, ec: Executor):
  @volatile var flushQ = scala.collection.mutable.LinkedHashSet.empty[Channel]

  scheduler.scheduleAtFixedRate(0 seconds, 1 millisecond)(() => flushEm())

  def apply(channel: Channel) =
    synchronized:
      flushQ += channel

  private def flushEm(): Unit =
    val toFlush = synchronized:
      val f = flushQ.take(100)
      flushQ --= f
      f
    // if flushQ.size > 500 then we're drowning, panic, log something, etc.
    toFlush.foreach(_.flush())
```

just an example

Reply:
and we play with the numbers. could be scheduleAtFixedRate(0, 500 milliseconds) and flushQ.take(5000), etc.

Reply:
i don't think it really matters, the clientEmit closure seems to recreate itself with some regularity.

Reply:
probably because i'm reconnecting..